### PR TITLE
Add parametric max and min function

### PR DESCRIPTION
**Pull request by haozhun**
Only long is implemented. Let me know if I can go ahead with the other 3 types. @martint

### CODE REVIEW

**Comment by haozhun** [comment]
Performance benchmark shows the following result. There doesn't seem like a significant change in performance. (`LongMaxAggregationSqlBenchmark`)

Parametric long max: 32.853 33.967 33.942 35.973 33.325
Long max (original): 34.677 32.374 36.535 34.082 32.958

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/LongMaxAggregationSqlBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import com.facebook.presto.testing.LocalQueryRunner;
+
+import static com.facebook.presto.benchmark.BenchmarkQueryRunner.createLocalQueryRunner;
+
+public class LongMaxAggregationSqlBenchmark
+        extends AbstractSqlBenchmark
+{
+    public LongMaxAggregationSqlBenchmark(LocalQueryRunner localQueryRunner)
+    {
+        super(localQueryRunner, "sql_varbinary_max", 40, 200, "select max(partkey) from lineitem");
```
Comment (cberner) [inline presto-benchmark/src/main/java/com/facebook/presto/benchmark/LongMaxAggregationSqlBenchmark.java]: This should say "sql_long_max"

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(
+                new ParameterMetadata(STATE),
+                new ParameterMetadata(INPUT_CHANNEL, type));
+    }
+
+    private static interface LongState
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: Name this LongMinMaxState, and move it to top level class since we'll want to reuse it for min()

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+        boolean isNull();
+
+        void setNull(boolean value);
+
+        long getLong();
+
+        void setLong(long value);
+    }
+
+    private static class FactoryForLong
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: Name this LongMinMaxStateFactory and move to top level class

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+            return new GroupedLongState(type, methodHandle);
+        }
+
+        @Override
+        public Class<? extends LongState> getGroupedStateClass()
+        {
+            return GroupedLongState.class;
+        }
+
+        public static class GroupedLongState
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: GroupLongMinMaxState

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+            }
+
+            @Override
+            public void setNull(boolean value)
+            {
+                nulls.set(getGroupId(), value);
+            }
+        }
+
+        public static class SingleLongState
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: SingleLongMinMaxState

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+                implements LongState
+        {
+            private final Type type;
+            private final MethodHandle methodHandle;
+            private long value;
+            private boolean nul;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: call this "isNull"

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+                this.type = type;
+                this.methodHandle = methodHandle;
+                this.value = 0;
+                this.nul = true;
+            }
+
+            @Override
+            public long getEstimatedSize()
+            {
+                return 16;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: Use ClassLayout.parseClass(SingleLongMinMaxState.class).instanceSize(), and cache it in a static field

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+
+            @Override
+            public void setNull(boolean value)
+            {
+                this.nul = value;
+            }
+        }
+    }
+
+    private static class StateSerializerForLong
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: Call this LongMinMaxStateSerializer

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+            implements AccumulatorStateSerializer<LongState>
+    {
+        @Override
+        public Type getSerializedType()
+        {
+            return VARCHAR;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: I think VARBINARY is more correct

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+        }
+
+        @Override
+        public void serialize(LongState state, BlockBuilder out)
+        {
+            SliceOutput sliceOutput = new DynamicSliceOutput((int) state.getEstimatedSize());
+            sliceOutput.writeLong(state.getLong());
+            sliceOutput.writeBoolean(state.isNull());
+            Slice slice = sliceOutput.slice();
+            out.writeBytes(slice, 0, slice.length());
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: Use VARBINARY.writeSlice

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+            out.closeEntry();
+        }
+
+        @Override
+        public void deserialize(Block block, int index, LongState state)
+        {
+            SliceInput input = block.getSlice(index, 0, block.getLength(index)).getInput();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: VARBINARY.getSlice

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+            long value = input.readLong();
+            state.setLong(value);
+            boolean nul = input.readBoolean();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: name this "isNull"

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java (hunk 2 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: Can't we use `com.facebook.presto.operator.aggregation.state.LongState` for this?

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: No, that doesn't have a method handle or type

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+            state.setNull(nul);
+        }
+    }
+
+    public static class ForLong
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: Move this to a top level class like MinMaxAggregationFunctions, so that we can reuse it for min()

**Comment by cberner** [comment]
Yep, looks like the right approach. Let me know when you've implemented it for the other types! Also, can you run the benchmarks in BenchmarkSuite?

**Comment by haozhun** [comment]
Please review again before I duplicate the files for the other 3 types.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregation.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    private static final Method LONG_INPUT_FUNCTION = method(LongMinMaxAggregationFunctions.class, "input", LongMinMaxState.class, long.class);
+    private static final Method LONG_COMBINE_FUNCTION = method(LongMinMaxAggregationFunctions.class, "combine", LongMinMaxState.class, LongMinMaxState.class);
+
+    public abstract String getName();
+    public abstract OperatorType getOperatorType();
+
+    @Override
+    public String getDescription()
+    {
+        return "Returns the maximum value of the argument";
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregation.java]: This needs to be abstract

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateSerializer.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        sliceOutput.writeBoolean(state.isNull());
+        Slice slice = sliceOutput.slice();
+        VARBINARY.writeSlice(out, slice);
+    }
+
+    @Override
+    public void deserialize(Block block, int index, LongMinMaxState state)
+    {
+        SliceInput input = VARBINARY.getSlice(block, index).getInput();
+        long value = input.readLong();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateSerializer.java]: inline this variable

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateSerializer.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        state.setLong(value);
+        boolean isNull = input.readBoolean();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateSerializer.java]: inline this one too

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public OperatorType getOperatorType()
+    {
+        return operatorType;
+    }
+
+    @Override
+    public Signature getSignature()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: I think this method can be implemented in the base class

**Comment by cberner** [comment]
Nice :) Assign back to me when you've finished with the other types

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java (hunk 17 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: I don't see how to do that without removing `static` from `SIGNATURE`.

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MaxAggregation.java]: I'd make SIGNATURE a member variable and create it in the constructor. Also, on that note I'd remove getName() and getOperatorType and add a constructor that takes them as arguments in the base class

**Comment by haozhun** [comment]
It now has `double`, `long`, and `Slice`. Adding `boolean` should be trivial, but `max` for `boolean` is currently an alias of `bool_or`.

**Comment by electrum** [comment]
Min should be boolean_and.

**Comment by haozhun** [comment]
Sorry. Let me rephrase the question: how to I add parametric max/min for boolean without breaking `boolean_and` and `boolean_or`?

A rather sub-optimal way is keep `boolean_and` and `boolean_or` there, but remove the `max`/`min` alias. I suppose there are other options. What are they?

**Comment by dain** [comment]
@electrum I don't think they can still be "aliases".  I think the parametric function needs to "own" all functions with the signature `T max(T)`.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateFactory.java [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+        }
+
+        @Override
+        public Type getType()
+        {
+            return type;
+        }
+
+        @Override
+        public MethodHandle getMethodHandle()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateFactory.java]: My comment about using LongState instead of LongMaxMinState seems to have been lost, and I can't find Chris' reply now, but I wanted to point out I think the method handle shouldn't be stored inside the state object. It should be bound into the method handle for the input function at specialization time, instead. 

The type is something that could be added to LongState, etc., and seems like a generally useful thing to do anyway. Once you do that, you can get rid of these extra state classes, factories, etc.

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateFactory.java]: Oh ya, I hadn't thought of binding a MethodHandle into another MethodHandle like that. I think we can actually bind both the MethodHandle and the Type in the specialize function, similar to how `ArrayEqualOperator` works. That would save a lot of code, since then I don't think we need any of these state objects.

Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/LongMinMaxStateFactory.java]: @martint Here is the previous discussion https://github.com/facebook/presto/pull/2446#discussion_r25887858

@cberner I don't understand the discussion. `ArrayEqualOperator` is not an aggregation function. And you can surely bind. But the situation is different here. `AggregationMetadata

**Comment by haozhun** [comment]
Depend on #2484.

Now that #2481 is done, everything is moved to the correct way. @martint 

Ready for review again.

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/LongMaxAggregationSqlBenchmark.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.benchmark;
+
+import com.facebook.presto.testing.LocalQueryRunner;
+
+import static com.facebook.presto.benchmark.BenchmarkQueryRunner.createLocalQueryRunner;
+
+public class LongMaxAggregationSqlBenchmark
```
Comment (cberner) [inline presto-benchmark/src/main/java/com/facebook/presto/benchmark/LongMaxAggregationSqlBenchmark.java]: Add this to the list of benchmarks in `BenchmarkSuite`

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 20]
```diff
@@ -316,11 +305,11 @@ public FunctionInfo load(SpecializedFunctionKey key)
                 .functions(MAP_EQUAL, MAP_NOT_EQUAL)
                 .functions(MAP_CONSTRUCTOR, MAP_CARDINALITY, MAP_SUBSCRIPT, MAP_TO_JSON, JSON_TO_MAP, MAP_KEYS, MAP_VALUES, MAP_AGG)
                 .function(IDENTITY_CAST)
-                .function(ARBITRARY_AGGREGATION)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: Add the arbitrary aggregation back. I assume this was a mistake?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregation.java [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+    public Signature getSignature()
+    {
+        return signature;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("E");
+        MethodHandle greaterThanMethodHandle = functionRegistry.resolveOperator(operatorType, ImmutableList.of(type, type)).getMethodHandle();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregation.java]: name this `methodHandle` since this is the base class

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/BooleanMinMaxAggregationFunctions.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+import static com.facebook.presto.spi.StandardErrorCode.INTERNAL_ERROR;
+
+public class BooleanMinMaxAggregationFunctions
+{
+    private BooleanMinMaxAggregationFunctions()
+    {
+    }
+
+    public static void input(final MethodHandle methodHandle, NullableBooleanState state, boolean value)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/BooleanMinMaxAggregationFunctions.java]: I don't think final is necessary here

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/BooleanMinMaxAggregationFunctions.java [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+            }
+        }
+        catch (Throwable t) {
+            Throwables.propagateIfInstanceOf(t, Error.class);
+            Throwables.propagateIfInstanceOf(t, PrestoException.class);
+            throw new PrestoException(INTERNAL_ERROR, t);
+        }
+    }
+
+    public static void combine(final MethodHandle methodHandle, NullableBooleanState state, NullableBooleanState otherState)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/BooleanMinMaxAggregationFunctions.java]: Same comment, do any of these need to be final?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/MinAggregation.java [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+
+    public MinAggregation()
+    {
+        super(NAME, OPERATOR_TYPE);
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Returns the maximum value of the argument";
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/MinAggregation.java]: this should say "minimum"

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregation.java [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+        MethodHandle combineFunction;
+        MethodHandle outputFunction;
+        Class<? extends AccumulatorState> stateInterface;
+
+        if (type.getJavaType() == long.class) {
+            stateFactory = compiler.generateStateFactory(NullableLongState.class, classLoader);
+            stateSerializer = new NullableLongStateSerializer();
+            stateInterface = NullableLongState.class;
+            inputFunction = LONG_INPUT_FUNCTION;
+            combineFunction = LONG_COMBINE_FUNCTION;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregation.java]: There's an optimization we can make here. The aggregation framework supports "intermediate input functions". These can be used instead of the combine function when the input and the intermediate input (the result of serializing a state object) are the same type.

So what you can do is use the input function as both the input, and the intermediate input function, and then remove the combine function.

You'll also need to add `type` as a parameter to the serializers, use that when serializing, and return it as the "intermediate type"

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/BooleanMinMaxAggregationFunctions.java (hunk 23 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/BooleanMinMaxAggregationFunctions.java]: Also see my comment above about re-using the input function instead of having this combine function

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/NullableSliceStateSerializer.java [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        if (state.isNull()) {
+            out.appendNull();
+        }
+        else {
+            SliceOutput sliceOutput = new DynamicSliceOutput((int) state.getEstimatedSize());
+            sliceOutput.writeInt(state.getSlice().length());
+            sliceOutput.writeBytes(state.getSlice());
+            Slice slice = sliceOutput.slice();
+            VARBINARY.writeSlice(out, slice);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/NullableSliceStateSerializer.java]: See my comment about including the `type` in the serializer. Then you can use it here

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/NullableSliceState.java [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+package com.facebook.presto.operator.aggregation.state;
+
+import io.airlift.slice.Slice;
+
+@AccumulatorStateMetadata(stateSerializerClass = NullableSliceStateSerializer.class)
+public interface NullableSliceState
+        extends AccumulatorState
+{
+    @InitialBooleanValue(true)
+    boolean isNull();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/NullableSliceState.java]: I don't think you need a boolean here, since you can just check if `getSlice()` returns `null`

**Comment by cberner** [comment]
Back to you for a couple more comments. Assign back to me when you're done with them

**Comment by haozhun** [comment]
Ready for review again. Also added test for Date (4-byte long type) and Array.

**Comment by cberner** [comment]
looks good

**Comment by dain** [comment]
👍
